### ISSUE TITLE
Wordpress app for iOS, Many false positives. Desperate help with rule exclusions.

### ISSUE THREAD

**Issue by Danrancan** [coreruleset/coreruleset#2392]
### Environment

* CRS version: v3.3.2
* Paranoia level setting: 1
* ModSecurity version: 1.0.2
* Web Server and version: Nginx Mainline 1.21.6
* Operating System and version: Ubuntu Server 20.04 (aarch64) for Raspberry pi

[x ] I have removed any personal data (email addresses, IP addresses,
    passwords, domain names) from any logs posted.

### Description

Hi,

I am very new to modsecurity, and have been struggling quite a bit with rule exclusions that are beyond the scope of examples shown in [this tutorial](https://www.linuxbabe.com/security/modsecurity-nginx-debian-ubuntu) from Linuxbabe.com. If this is the wrong place to make this post I apologize in advance. I am running a Wordpress LEMP Server, and my issue is specific to getting the iOS Wordpress app working properly, without modsecurity flagging posts from the iOS Wordpress app as attacks. Also, note, I am using the OSWAP ruleset with modsecurity.

The core issues I am having with my wordpress app, are that when turning plugin's on and off from the iOS app, modsecurity denies that action. The second (and more important) issue I am having with the iOS app is making posts from the iOS app. Specifically I am having trouble making posts that have a single picture, or a picture gallery in the iOS app.

Initially, the iOS app would not let me make any posts whatsoever, including plain text posts. 
As of now, I have managed to create a rule exclusion to allow plain text posts, however, that is as far as I have gotten, and any other attempts to unblock (with rule exclusions) picture uploads and gallery posts containing photos, have been a failure. Furthermore any attempts to unblock plugin use and activation from the iOS Wordpress app have also been a failure.

I have been tinkering with this for about 2 months now, and am resorting to posting on here in a desperate attempt for a pro to look at my logs and possibly help me create the proper rule exclusions to get my iOS app fully working.

#### **Problem 1**) Solved

To begin, these are the logs that were created when trying to make a TEXT BASED POST on the Wordpress iOS app:
[https://pastebin.com/N0UugaAc](https://pastebin.com/N0UugaAc)

After looking at these logs and grasping a mild understanding from linuxbabe.com's tutorial, i created the following rule in the `REQUEST-900-EXCLUSION-RULES-BEFORE-CRS.conf` file:

```
# Allow iOS app to create and save TEXT BASED posts. This rule does not allow plugins to work yet. WORKING!
SecRule REQUEST_URI "@streq /xmlrpc.php" "id:1030,phase:1,ctl:ruleRemoveById=941180"
```
This rule alone successfully allows iOS to make text based posts, but does not allow posting photos or uploading photos. Problem 1 solved.

#### **Problem 2**) Unsolved

The next problem I am having is trying to upload photos (multiple as well as single) and make a photo based post from the iOS app. These are the logs that were created when trying to make a TEXT BASED POST on the Wordpress iOS app:

[https://pastebin.com/nN7h5stb](https://pastebin.com/nN7h5stb)

I have tried the following rules exclusions to allow multi-photo blog posts in wordpress, and **none have worked** thus far.

```
# Allow iOS Wordpress app to create and save picture based posts.
SecRule REQUEST_URI "@streq /xmlrpc.php" "id:1031,phase:1,ctl:ruleRemoveById=941100"
SecRule REQUEST_URI "@streq /xmlrpc.php" "id:1033,phase:1,ctl:ruleRemoveByID=941160"
SecRule REQUEST_URI "@streq /xmlrpc.php" "id:1034,phase:1,ctl:ruleRemoveByID=932140"
```

Could someone please do me a big great favor and take a look at the above logs, and help me create a proper rule exclusion for posting photos with the Wordpress iOS app?

#### **Problem 3**) Unsolved

The Final exclusion I need is to allow the wordpress iOS app to turn on and off plugins. Currently plugins are on. Below is the log posted when trying to turn a plugin off....

[https://pastebin.com/jsbC2HBd](https://pastebin.com/jsbC2HBd)

I have tried the following rules exclusions to plugin activation and deactivation (Specifically the Flamingo plugin) from within the iOS Wordpress App, and **none have worked** thus far.

```
SecRule REQUEST_URI "@within /wp-json/wp/v2/plugins/flamingo/flamingo" "id:1040,phase:2,ctl:ruleRemoveById=1642837425"
```
Could someone, again, please do me a big great favor and take a look at the above logs, and help me create a proper rule exclusion for activating and deactivating plugins in the Wordpress iOS app?

Super sorry for the long post, and again, sorry if this has been posted in the wrong place. I'm just desperate at this point.
**Thank you supremely** to any and all people who can help me out here with the proper exclusions. 

Thank you thank you thank you!
Dan

**Comment by RedXanadu** [comment]
Hi @Danrancan,

Firstly, have you seen our **official documentation**? We have a detailed section covering false positives and tuning via rule exclusions. You can find it [here](https://coreruleset.org/docs/configuring/false_positives_tuning/). The eight rule exclusion scenarios and examples may be particularly useful!

Secondly, have you tried using **our WordPress rule exclusion package**? For several popular web applications, including WordPress, we have pre-made packages of rule exclusions to help deal with the most common false positives / false alarms. You can find more information on how to use our rule exclusion packages [here](https://coreruleset.org/docs/configuring/false_positives_tuning/#rule-exclusion-packages).

Next, some specific comments about your rule exclusions. It's always a good idea to explicitly use the `pass` action with this type of rule exclusion. This makes it clear that the rule should never block any transactions. Secondly, you almost certainly want to use the `nolog` action. This prevents your rule exclusions from generating log entries of their own when they match, which clutters the log files and makes it harder to find and deal with genuine false positives. Try something like:

```apache
# Allow iOS app to create and save TEXT BASED posts. This rule does not allow plugins to work yet. WORKING!
SecRule REQUEST_URI "@streq /xmlrpc.php" \
    "id:1030,\
    phase:1,\
    pass,\
    nolog,\
    ctl:ruleRemoveById=941180"
```

The line breaks and indentation aren't strictly necessary, but you may find rules easier to read when they're laid out like this, with one action per line.

Lastly, and perhaps most importantly, it looks like your **inbound anomaly score threshold** is set to 5. _That's very low!_ Getting the inbound threshold down to 5 is a great end goal to have, but it sounds like starting out at 5 is making your life unnecessarily difficult.

There's a school of thought that advocates starting with a _very_ high threshold (so high that, in practise, nothing will ever be blocked: say, 10000), using the web app through the WAF, and then inspecting the log files and creating rule exclusions (making sure to _only_ create exclusions based on known, good traffic: maybe filtering the log lines for your own IP address, if your service is publicly accessible). Then the threshold can be brought down step by step over time, as your confidence in the solution grows and the rule exclusions become more refined. You can find an excellent tutorial written by our project co-leader, who describes this process [here](https://www.netnea.com/cms/apache-tutorial-8_handling-false-positives-modsecurity-core-rule-set/).

The log entries regarding your 🦩 app are throwing an error on Pastebin, so I can't offer any advice there.

Please try following the advice above, especially activating the WordPress rule exclusion package. If you're still having problems dealing with specific false positives and writing rule exclusions then please do reply again here. Let us know how you get on.

Good luck!

**Comment by RedXanadu** [comment]
Hi @Danrancan,

I'm going to close this issue now, but feel free to reply again here if you run into any further issues or if you have any more questions.
